stl/inc/numeric

```diff
@@ -70,31 +70,40 @@ inline constexpr bool _Plus_on_arithmetic_ranges_reduction_v = false;
 #endif // _STD_VECTORIZE_WITH_FLOAT_CONTROL
 
 template <class _InIt, class _Ty, class _BinOp>
-_NODISCARD _Ty reduce(const _InIt _First, const _InIt _Last, _Ty _Val, _BinOp _Reduce_op) {
+#ifdef __cpp_lib_is_constant_evaluated
+_NODISCARD _CONSTEXPR20 _Ty reduce(
+#else // ^^^ __cpp_lib_is_constant_evaluated / !__cpp_lib_is_constant_evaluated vvv
+_NODISCARD _Ty reduce(
+#endif // __cpp_lib_is_constant_evaluated
+    const _InIt _First, const _InIt _Last, _Ty _Val, _BinOp _Reduce_op) {
     // return commutative and associative reduction of _Val and [_First, _Last), using _Reduce_op
     _Adl_verify_range(_First, _Last);
     auto _UFirst      = _Get_unwrapped(_First);
     const auto _ULast = _Get_unwrapped(_Last);
     if constexpr (_Plus_on_arithmetic_ranges_reduction_v<_Unwrapped_t<const _InIt&>, _Ty, _BinOp>) {

```

Comment:
Unfortunately this is suboptimal because it's still trying to compile the "generic" case even when the special case is engaged, which was previously stopped by `if constexpr`. That loop is small enough that I think it's worth duplicating, and if it's big enough it's worth factoring into another function.

Reply:
I dont believe you are correct here. There are 2 cases, first `!_Plus_on_arithmetic_ranges_reduction_v(...)`

There the `if constexpr` clause is removed and we are left with the simple loop, which is the optimal code. In the other case (`_Plus_on_arithmetic_ranges_reduction_v`) we are left with 
```cpp
#ifdef __cpp_lib_is_constant_evaluated
    if (!_STD is_constant_evaluated())
#endif // __cpp_lib_is_constant_evaluated
    {
        (void) _Reduce_op; // TRANSITION, VSO-486357
        return _Reduce_plus_arithmetic_ranges(_UFirst, _ULast, _Val);
    }
    for (; _UFirst != _ULast; ++_UFirst) {
        _Val = _Reduce_op(_STD move(_Val), *_UFirst); // Requirement missing from N4713
    }

    return _Val;
```

AFAIK the `if(is_constant_evaluated)` branch is a `classic` if-branch where the else is still processed by the compiler. Consequently, we cannot avoid that the compiler needs to process the simple loop, even in the non-constexpr case.

Reply:
To be more precise, It is suboptimal in comparison to the previous code, but is the optimal solution given the requirements of `constexpr`.

Reply:
https://godbolt.org/z/BYL3Go

Reply:
I test on clang this code, and it's working fine.
But if requires I can change this.

Reply:
@Neargye I believe the comment from @BillyONeal is geared towards troughput not correctness. Previously the compiler would discard one of the branches when processing the function. Now it always has to process the loop, which is a -most likely non mesurable- regression in throughput. 

However, given the constexpr requirement I cannot see a way around that given that the `if (is_constant_evaluated())` doesnt discard branches

Reply:
>-most likely non mesurable-

It instantiates ~2 extra function templates (_STD move and _Reduce_op). Admittedly in this case that probably isn't a huge deal; this is more an "in general, we are willing to jump through hoops to instantiate less stuff". This is one of the things that you have to have constant vigilance; otherwise these here-and-there places add up over time.